File: python/sort_.py
#!/usr/bin/python3

import argparse
import re
import sys
from copy import deepcopy as dcopy
from itertools import chain
from typing import Literal

def preprocess(
        file_name, 
        unique: bool = False, 
        add_rand_hash: bool = False,
        radix: bool = False
    ):
    """
    Processing in the .txt file

    Parameters
    ----------
    file_name : str, file_path like
        .txt file to be ingested
    unique : bool, optional
        Based on -u flag arg, by default False
    add_rand_hash : bool, optional
        Based on the -R flag arg, by default False
    radix : bool, optional
        Based on the sort_type arg, by default False

    Returns
    -------
    list, list, dict, int
        Four outputs:
            words/words_set: list
                list of individual words to be sorted, can be unique or
                non-unique based on unique arg
            idx_to_sort: list
                only used for -R sorting, sorts the indicies then maps 
                back to the words at the end
            idx_word_map: dict
                only used for -R sorting, dictionary to map index list 
                back to the words for output.
             max_len: int
                longest word in the words list, used for the radix sort
    """
    # make placeholder vars (last two only used for -R sorting)
    words = []
    max_len = 0
    idx_to_sort = []
    idx_word_map = {}

    # read in .txt file
    with open(file_name, 'r', encoding='utf-8-sig') as file:
        # use regex to find valid word strings, add the line if not empty
        for line in file:
            add_line = re.findall(r"[A-Za-z]+", line)
            if add_line:
                word_lengths = [len(word) for word in add_line]
                # keep running track of max_len
                max_len = max(max_len, max(word_lengths))
                [words.append(word) for word in add_line]
    
    # sorting output based on input args
    if add_rand_hash or unique:
        words_set = list(set(words))
        # make mapping index list to be sorted for -R sorting
        if add_rand_hash:
            if radix:
                hash_idx = [f"{i}" for i in range(len(words_set))]
            else:
                hash_idx = [i for i in range(len(words_set))]
            # word_idx_map for converting words to indicies to sort
            word_idx_map = {word:i for word, i in zip(words_set, hash_idx)}
            # idx_word_map to map those indicies back to words
            idx_word_map = {i:word for i, word in zip(hash_idx, words_set)}
            # list that will actually be sorted during -R sorting
            idx_to_sort = list(map(word_idx_map.get, words))
            words = words_set if unique else words
            return words, idx_to_sort, idx_word_map, max_len
        # for just unique sorts
        else:
            return words_set, idx_to_sort, idx_word_map, max_len
    # non-unique and non -R sorting
    else:
        return words, idx_to_sort, idx_word_map, max_len
    

def quick_sort(words):
    """
    Standard quick sort algorithm

    Parameters
    ----------
    words : list
        list of words to be sorted

    Returns
    -------
    list
        sorted list
    """
    # base case
    if len(words) <= 1:
        return words 
    else:
        # choose half way pivot
        pivot_position = len(words) // 2
        pivot = words[pivot_position]

        # sort based on values related to pivot
        left = [w for w in words if w < pivot]
        middle = [w for w in words if w == pivot]
        right = [w for w in words if w > pivot]

        # sort partitioned lists until base case
        return quick_sort(left) + middle + quick_sort(right)

def merge_sort(words):
    """
    Merge sort algorithm

    Parameters
    ----------
    words : list
        list of words to be sorted

    Returns
    -------
    list
        sorted list
    """
    # base case
    if len(words) <= 1:
        return words 
    else:
        # break list in two until all element in list of len == 1
        mid = len(words) // 2
        left = merge_sort(words[:mid])
        right = merge_sort(words[mid:])

        # merge partitioned lists to sort
        return merge(left, right)

def merge(left, right):
    """
    Merge sort sub routine

    Parameters
    ----------
    left : list
        left partition from merge function
    right : list   
        right partition from merge function

    Returns
    -------
    list   
        sorted list
    """
    output = []
    i = j = 0

    # while still have elements to compare in both lists
    while i < len(left) and j < len(right):
        # append to output stack based on vals
        if left[i] < right[j]:
            output.append(left[i])
            i += 1
        else:
            output.append(right[j])
            j += 1

    # put the rest of the values of the list we didn't get through
    # at the top of the stack (this is fine because these elements
    # are all greater than the last element added)
    output += left[i:]
    output += right[j:]

    return output

def heap_sort(words):
    """
    Heap sort algorithm

    Parameters
    ----------
    words : list
        list to be sorted

    Returns
    ----------
    None: sorts the list in place
    """
    n = len(words)

    # build max heap
    for i in range(n // 2 - 1, -1, -1):
        heapify(words, n, i)
    
    # take out inidividual elements from heap
    for i in range(n - 1, 0, -1):
        words[i], words[0] = words[0], words[i]
        heapify(words, i, 0)

def heapify(words, n, i):
    """
    Heap sort sub routine

    Parameters
    ----------
    words : list
        list to be heapified
    n : int
        length of heap to modify
    i : int
        index of initial root of heap
    """
    # big is root of heap, left/right are children indicies
    big = i
    left = 2 * i + 1
    right = 2 * i + 2

    # update if children larger than root
    if left < n and words[left] > words[big]:
        big = left 
    
    # check both children
    if right < n and words[right] > words[big]:
        big = right 

    # orig root not largest
    if big != i:
        # swap these vals
        words[i], words[big] = words[big], words[i]
        # heapify subtree
        heapify(words, n, big)


def radix_sort(words, longest_word):
    """
    Radix sort algorithm

    Parameters
    ----------
    words : list
        list to be sorted
    longest_word : int
        length of longest word in the list

    Returns
    -------
    list
        sorted list
    """
    # these are non-letters in the middle of the chr range from 'A' to 'z'
    bad_set = {i for i in range(91,97)}
    # generates a string from 'A' to 'z' in lexicographical order
    order = "".join([chr(i) for i in range(65,123) if i not in bad_set])
    # base empty table to store the stacks/queues in the sort
    std_radix = [[] for char in order]
    max_iter = longest_word
    # sets which index letter we're evaluating for every word
    level = max_iter

    # do until first characters are encountered
    while level > 0:
        # make a new table to move values to
        new_radix = dcopy(std_radix)

        # for first time through
        if level == max_iter:
            for word in words:
                # count shorter words as being right padded with the 'A' 
                # character, similar to zero padding integers
                if len(word) < max_iter:
                    new_radix[0].append(word)
                # move the word to the appropriate bin
                else:
                    letter = word[-1]
                    idx = map_val(letter)
                    # add to stack in that bin
                    new_radix[idx].append(word)
            # for use in next iteration
            old_radix = dcopy(new_radix)
        # for all other iterations
        else:
            # for each character bin
            for j in range(len(order)):
                # if the bin is not empty
                while old_radix[j]:
                    # treat old stack as queue now
                    temp = old_radix[j].pop(0)
                    # same 'A' padding logic used in first step
                    if len(temp) < level:
                        idx = 0
                    # if word has valid letters at this length, calculcate
                    # where that word should go next
                    else:
                        letter = temp[level - 1]
                        idx = map_val(letter)
                    # move word to next appropriate bin stack
                    new_radix[idx].append(temp)
            # store for next use
            old_radix = dcopy(new_radix)
        level -= 1
    # just output standard list
    return list(chain(*new_radix))


def map_val(letter):
    """
    Mapping alphabetical string to ord value

    Parameters
    ----------
    letter : str
        ascii letter

    Returns
    -------
    int
        index in string of alphabetical characters used in radix sort
    """
    num_val = ord(letter)
    if num_val < 97:
        return num_val - 65
    else:
        return num_val - 71

def sort(
        words, 
        sort_type: Literal['radix','merge','quick','heap'], 
        max_len
    ):
    """
    General sort director

    Parameters
    ----------
    words : list
        list to be sorted
    sort_type : Literal[radix, merge, quick, heap]
        type of sort to conduct
    max_len : int
        length of longest word, used only for radix sort

    Returns
    -------
    list
        sorted list
    """
    if sort_type == "quick":
        return quick_sort(words)
    elif sort_type == "merge":
        return merge_sort(words)
    elif sort_type == "heap":
        heap_sort(words)
        return words
    else:
        return radix_sort(words, max_len)
    

def create_parser():
    """
    Command line argument parser

    Returns
    -------
    argparse.ArgumentParser
        Namespace for command line argumetns
    """
    parser = argparse.ArgumentParser(description="Sort App")
    parser.add_argument("filename", help=".txt file to be sorted")
    parser.add_argument(
        "sort_type",
        choices=["radix", "merge", "quick", "heap"],
        default="quick",
        help="sort method"
    )
    parser.add_argument(
        "-u",
        "--uniq",
        help="sort unique words", 
        action="store_true"
    )
    parser.add_argument(
        '-d',
        '--desc',
        help="sort in descending order",
        action="store_true"
    )
    parser.add_argument(
        "-R",
        "-random-sort",
        help="sort by a random order",
        action="store_true"
    )
    return parser


def main():
    """
    Main function for script
    """
    # read in arguments
    parser = create_parser()
    args = parser.parse_args()

    # ingest text file, create helper variables based on args
    ingest = preprocess(
        args.filename, 
        unique=args.uniq,
        add_rand_hash=args.R,
        radix=True if args.sort_type=="radix" else False 
    )
    # unpack output from preprocess function
    words, idx_to_sort, idx_word_map, max_len = [i for i in ingest]

    # random sort
    if args.R:
        sorted_idx = sort(idx_to_sort, args.sort_type, max_len)
        sorted_words = list(map(idx_word_map.get, sorted_idx))
    # sort normal
    else:
        sorted_words = sort(words, args.sort_type, max_len)
    # reverse the sort if they ask for it
    if args.desc:
        sorted_words.reverse()
    # output all unless piped into head function
    for word in sorted_words:
        try: 
            print(word)
        except BrokenPipeError:
            sys.exit(0)

if __name__ == '__main__':
    main()

File: c/sort_.c

#include <stdio.h>
#include <string.h>
#include <sys/stat.h>
#include <stdbool.h>
#include <stdlib.h>
#include <ctype.h>
#include <time.h>
#include "hash_chr.h"

#define MAX_WORDS 100000
#define MAX_WORD_LENGTH 100

char word[MAX_WORD_LENGTH];
char *words[MAX_WORDS];
int wordCount = 0;

// printing arrays
void printArray(char **arr, int size) {
    for (int i = 0; i < size; i++) {
        printf("%s\n", arr[i]);
    }
}

// safer version of strcpy, checks memory first
void sstrcpy(char** arr, char *new_item, int index) {
    arr[index] = malloc(strlen(new_item) + 1);
    if (arr[index] == NULL) {
        perror("Memory allocation failed");
        return;
    }
    strcpy(arr[index], new_item);
}

// grabs alphabetic characters only
void cleanWord(char *word) {
    int i = 0, j = 0;
    while (word[i]) {
        if (isalpha(word[i])) {
            word[j++] = word[i];
        }
        i++;
    }
    word[j] = '\0';
}

// reading in txt file
void readFile(const char *filename) {
    // open file
    FILE *file;
    file = fopen(filename, "r");

    // Read words from the file one by one (separates words by white space)
    while (fscanf(file, "%s", word) == 1) { // read each word up to 99 characters long
        if (wordCount >= MAX_WORDS) {
            printf("Too many words in file.\n");
            break;
        }
        cleanWord(word);
        // only add words of alphabetic letters, the first character
        // of words without them will be '\0' after the cleanWord
        // function, so filter those out
        if (word[0] == '\0') {
            continue;
        } else {
            // dynamically allocate memory for each word and store it
            words[wordCount] = malloc(strlen(word) + 1);
            if (words[wordCount] == NULL) {
                perror("Memory alloction failed");
                return;
            }
            strcpy(words[wordCount], word);
            wordCount++;
        }
    }
    fclose(file);
}

// preprocess the file
long getFileSize(const char *filename) {
    FILE *file = fopen(filename, "rb");
    if (file == NULL) {
        perror("Error opening file");
        return -1;
    }
    // move file pointer to end of the file
    fseek(file, 0, SEEK_END);
    // get current position of the file pointer, == size of file
    long size = ftell(file);
    // close the file
    fclose(file);

    return size;
}

/// quick sort stuff
// swapping two strings in array
void swap(char** a, char** b) {
    char* temp = *a;
    *a = *b;
    *b = temp;
}
// parition subrouting used for quick sort
int partition(char** arr, int low, int high) {
    // pointer to initial pivot, smaller element
    char* pivot = arr[high];
    int i = low - 1;
    // loop through to end, swapping elements along
    // the way if less than pivot
    for (int j = low; j < high; j++) {
        if (strcmp(arr[j], pivot) < 0) {
            i++;
            swap(&arr[i], &arr[j]);
        }
    }
    // move pivot to where it should be to partition
    // the list where everything < pivot is left of it
    // and everything > pivot is right of it
    swap(&arr[i + 1], &arr[high]);
    return (i + 1);
}
// main in place quick sort function via partition method
void quickSort(char** arr, int low, int high) {
    if (low < high) {
        // partition array into two sublists based on
        // last element
        int part = partition(arr, low, high);
        // recursively sort sublists until sorted
        quickSort(arr, low, part - 1);
        quickSort(arr, part + 1, high);
    }
}
/// end quick sort stuff

/// merge sort stuff
// merge two halves
void merge(char** arr, int left, int mid, int right) {
    int i, j, k;
    int n1 = mid - left + 1; // left sub array
    int n2 = right - mid; // right sub array
    // temp arrays for subs
    char *L[n1], *R[n2];
    // copy words into sub temp arrays
    for (i = 0; i < n1; i++) {
        sstrcpy(L, arr[left + i], left + i);
    }
    for (j = 0; j < n2; j++) {
        sstrcpy(R, arr[mid + 1 + j], mid + 1 + j);
    }

    // merge temp arrays back into arr[left..right]
    i = 0;
    j = 0;
    k = left;
    while (i < n1 && j < n2) {
        if (strcmp(L[i], R[j]) <= 0) {
            // copy L[i] into arr[k]
            sstrcpy(arr, L[i], k);
            i++;
        } else {
            // copy R[j] into arr[k]
            sstrcpy(arr, R[j], k);
            j++;
        }
        k++;
    }

    // copy remaining elements of L[] and R[] if any
    while (i < n1) {
        sstrcpy(arr, L[i], k);
        i++;
        k++;
    }
    while (j < n2) {
        sstrcpy(arr, R[j], k);
        j++;
        k++;
    }
}
// main merge sort func
void mergeSort(char** arr, int left, int right) {
    if (left < right) {
        // find middle point
        int mid = left + (right - left) / 2;

        // sort left and right of mid
        mergeSort(arr, left, mid);
        mergeSort(arr, mid + 1, right);

        // merge sorted sub arrays
        merge(arr, left, mid, right);
    }
}
//// end merge sort stuff

// displaying help function
void print_help() {
    printf("Usage sort_ [-h] [-u] [-d] [-R] filename {merge,quick}\n");
    printf("(heap, radix coming soon)\n");
    printf("Sort App\n\n");
    printf("positional arguments:\n");
    printf("  filename\t\t\t.txt file to be sorted\n");
    printf("  {merge,quick}\t\t\tsort method\n\n");
    printf("options:\n");
    printf("  -h, --help\t\t\tshow this help message and exit\n");
    printf("  -u, --uniq\t\t\tshow unique words\n");
    printf("  -d, --desc\t\t\tsort in descending order\n");
    printf("  -R, -random-sort\t\t(also coming soon) sort by a random order\n");
}

int main(int argc, char *argv[]) {
    char *filename;
    // set the defaults
    char *sort_type = "quick";
    bool uniq = false;
    bool desc = false;
    bool rand = false;
    
    if (argc > 1){
        // display help option if first argument
        if (strcmp(argv[1], "-h") == 0 || strcmp(argv[1], "--help") == 0) {
            print_help();
            return 0;
        } else {
            filename = argv[1];
            struct stat buffer;

            // verify the file exists
            if (stat(filename, &buffer) != 0) {
                printf("no file at %s\n", filename);
                return 1;
            }
        }
    }
    // if there are more arguments passed
    if (argc > 2) {
        // parse the args
        for (int i=2; i < argc; i++) {
            if (strcmp(argv[i], "-u") == 0 || strcmp(argv[i], "--uniq") == 0) {
                uniq = true;
            }
            if (strcmp(argv[i], "-d") == 0 || strcmp(argv[i], "--desc") == 0) {
                desc = true;
            }
            if (strcmp(argv[i], "-R") == 0 || strcmp(argv[i], "-random-sort") == 0) {
                rand = true;
            }
            if (strcmp(argv[i], "quick") == 0) {
                sort_type = "quick";
            }
        }
    }
    if (filename == NULL) {
        printf("no filename passed\n");
        return 1;
    } 

    printf("sort tool ready to run w/ following params\n");
    printf("filename:\t%s\n", filename);
    printf("sort_type:\t%s\n", sort_type);
    printf("uniq:\t\t%d\n", uniq);
    printf("desc\t\t%d\n", desc);
    printf("rand\t\t%d\n", rand);

    // read in words from the file
    readFile(filename);

    // final_words variables
    char *final_words[MAX_WORDS];
    int final_count = 0;

    // make unique group if passed, else just copy words over
    if (uniq) {
        printf("making unique word set\n");
        hash_table *words_set = hash_init(MAX_WORDS);

        for (int i=0; i < wordCount; i++) {
            char *word = words[i];
            hash_insert(words_set, word);
        }
        printf("set successfully made\n");

        // iterate through hash table to get unique words to sort
        int curr_size = words_set->num_entries;
        int index;
        hash_entry *curr;
        char *curr_word;
        printf("num_entries:\t%d\n", curr_size);
        for (int i=0; i < curr_size; i++) {
            index = words_set->taken[i];
            curr = words_set->entries[index];
            curr_word = curr->value;
            sstrcpy(final_words, curr_word, i);
            final_count++;
        }
        hash_destroy(words_set);
        printf("words read in, hash table destroyed");
    } else {
        for (int i=0; i < wordCount; i++) {
            sstrcpy(final_words, words[i], i);
            final_count++;
        }
    }

    printf("\nSorted words in the file:\n");
    printf("final_count\t%d\n", final_count);

    // choose sort type
    if (strcmp(sort_type, "quick") == 0){
        quickSort(final_words, 0, final_count - 1);
    } else if (strcmp(sort_type, "merge") == 0) {
        mergeSort(final_words, 0, final_count - 1);
    }
    
    // if descending order is requested
    if (desc) {
        int end = final_count / 2 - 1;
        int j = final_count - 1;
        for (int i=0; i < end; i++) {
            swap(&final_words[i], &final_words[j]);
            j--;
        }
    }
    printArray(final_words, final_count);
}
